### ISSUE TITLE
dartdevc: toString() crashes on non-Dart object

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#35952]
A customer reports the following stack trace when trying to interpolate a `GwtException` from a GWT  app in the page (paths anonymized).

Clearly it should not crash. It seems like the dart.str helper should be hardened.
Opened #35951 to answer the separate question of _what_ it should do.

```
TypeError: obj[$toString] is not a function
    at Object.dart_library.library.dart.str (https://.../resources/dart_sdk.js)
    at Object.dart_library.library.expression9.temporaryFunction82436245992617 (http://dev_compiler_debugger/last.js:21:17)
    at eval (eval at _onException (https://.../resources/packages/.../exceptions.js), <anonymous>:1:71)
    at eval (eval at _onException (https://.../resources/packages/.../exceptions.js), <anonymous>:1:324)
    at dart_library.library.exception_handler.ConsoleExceptionHandler.new._onException (https://.../resources/packages/.../exceptions.js)
    at _onException.next (<anonymous>)
    at runBody (.../resources/dart_sdk.js)
```

There is a thread with the title 'Check if object is "NativeJavaScriptObject"' with more details.

**Comment by rakudrama** [comment]
Note that `Object.create(null).toString()` fails with a TypeError.

**Comment by jmesserly** [comment]
Yeah, it's an object without `Object.prototype` in the proto chain. This is an issue for all members on dart:core Object. The question is to what extent we want to slow them down.

> Clearly it should not crash. It seems like the dart.str helper should be hardened.
> Note that `Object.create(null).toString()` fails with a TypeError.

Yeah, JavaScript `toString()` behavior is identical: it throws a TypeError. It's not clear to me we need to change this behavior (and I wouldn't describe it as a "crash"--it is an unsupported operation on those objects). We may want to have friendlier behavior than JS does, especially for `toString()` which is used for error reporting. (Conceptually it's nice if `toString()` never throws... but there's no way to guarantee that in Dart or JS.)

I'll think more on this in the other bug. I would be hesitant to implement the other Object members, especially `hashCode` and `==` as that may incur too much performance cost.

Regarding the stack trace, it is interesting that "dev_compiler_debugger", "eval" and "temporaryFunction82436245992617" is on the stack. This looks like it is using the Dart debugger extension code somehow. @alan-knight does that stack trace look familiar at all? I found the corresponding internal code for "ConsoleExceptionHandler" (can send you link, it's easy to find though) but I don't understand where the "eval" and "dev_compiler_debugger" stuff is coming from.

**Comment by jmesserly** [comment]
(marked blocked, as we'll need to address https://github.com/dart-lang/sdk/issues/35951 first.)
